mobilecoind/src/db_crypto.rs

```diff
@@ -0,0 +1,186 @@
+    MissingEncryptionKey,
+
+    #[fail(display = "Database encryption not enabled")]
+    EncryptionNotEnabled,
+
+    #[fail(display = "Encryption key already set")]
+    EncryptionKeyAlreadySet,
+
+    #[fail(display = "Invalid encryption key length")]
+    EncryptionInvalidKeyLength,
+
+    #[fail(display = "AEAD error: {}", _0)]
+    Aead(AeadError),
+}
+
+impl From<AeadError> for DbCryptoError {
+    fn from(src: AeadError) -> Self {
+        Self::Aead(src)
+    }
+}
+
+/// Interface for database encryption support.
+pub trait DbCryptoProvider: Clone {
+    /// Sets the password to use.
+    fn set_password(&self, password: &[u8]) -> Result<(), DbCryptoError>;
+
+    /// Clear the password.
+    fn clear_password(&self) -> Result<(), DbCryptoError>;
+
+    /// Whether a password is needed or not (allows us to tell if we're running with an encryption
+    /// backend or without).
+    fn requires_password(&self) -> bool;
+
+    /// Encrypt data.
+    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError>;
+
+    /// Decrypt data.
+    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError>;
+}
+
+/// A no-op crypto provider, used when encryption is disabled.
+#[derive(Clone, Default)]
+pub struct NoDbCryptoProvider;
+
+impl DbCryptoProvider for NoDbCryptoProvider {
+    /// Sets the password to use.
+    fn set_password(&self, _password: &[u8]) -> Result<(), DbCryptoError> {
+        Err(DbCryptoError::EncryptionNotEnabled)
+    }
+
+    /// Clear the password.
+    fn clear_password(&self) -> Result<(), DbCryptoError> {
+        Err(DbCryptoError::EncryptionNotEnabled)
+    }
+
+    /// Whether a password is needed or not (allows us to tell if we're running with an encryption
+    /// backend or without).
+    fn requires_password(&self) -> bool {
+        false
+    }
+
+    /// Encrypt data.
+    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+
+    /// Decrypt data.
+    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+}
+
+/// Aes crypto provider
+#[derive(Clone)]
+pub struct AesDbCryptoProvider {
+    encryption_key: Arc<Mutex<Option<Vec<u8>>>>,
+}
+
+impl Default for AesDbCryptoProvider {
+    fn default() -> Self {
+        Self {
+            encryption_key: Arc::new(Mutex::new(None)),
+        }
+    }
+}
+
+impl DbCryptoProvider for AesDbCryptoProvider {
+    /// Sets the password to use.
+    /// The password needs to be 32 bytes - the intended purpose is for this to be a hash of the
+    /// password and not the actual password the user typed.
+    fn set_password(&self, password: &[u8]) -> Result<(), DbCryptoError> {
+        let mut encryption_key = self.encryption_key.lock().expect("muted poisoned");
+        if encryption_key.is_some() {
+            Err(DbCryptoError::EncryptionKeyAlreadySet)
+        } else if password.len() != 32 {
```

Comment:
perhaps this 32 should be pulled out as a constant as a part of the AesDbCryptoProvider?